File: 15-String/java/06-RepeatedStringMatch.java
class Solution {
    public int repeatedStringMatch(String a, String b) {
        boolean[] charInA = new boolean[26];
        for (char c : a.toCharArray()) {
            charInA[c - 'a'] = true;
        }
        for (char c : b.toCharArray()) {
            if (!charInA[c - 'a']) return -1;
        }

        int minReps = (b.length() + a.length() - 1) / a.length();

        StringBuilder repeated = new StringBuilder();
        for (int i = 0; i < minReps; i++) {
            repeated.append(a);
        }

        if (repeated.toString().contains(b)) {
            return minReps;
        }

        repeated.append(a);
        if (repeated.toString().contains(b)) {
            return minReps + 1;
        }

        return -1;
    }
}

File: 15-String/python/06_repeated_string_match.py
class Solution:
    def repeatedStringMatch(self, a: str, b: str) -> int:
        set_a = set(a)
        for c in b:
            if c not in set_a:
                return -1

        min_reps = (len(b) + len(a) - 1) // len(a)
        repeated = a * min_reps

        if b in repeated:
            return min_reps

        if b in repeated + a:
            return min_reps + 1

        return -1
